go-apps/meep-loc-serv/server/loc-serv.go

```diff
@@ -1204,49 +1589,599 @@ func zonalTrafficSubGet(w http.ResponseWriter, r *http.Request) {
+			http.Error(w, "Invalid Mandatory EnteringLeavingCriteria parameter value", http.StatusBadRequest)
+			return
+		}
+	}
+	if areaCircleSub.Frequency == 0 {
+		log.Error("Mandatory Frequency parameter not present")
+		http.Error(w, "Mandatory Frequency parameter not present", http.StatusBadRequest)
+		return
+	}
+	/*
+	   if areaCircleSub.CheckImmediate == nil {
+	           log.Error("Mandatory CheckImmediate parameter not present")
+	           http.Error(w, "Mandatory CheckImmediate parameter not present", http.StatusBadRequest)
+	           return
+	   }
+	*/
+	/*
+		if areaCircleSub.TrackingAccuracy == 0 {
+			log.Error("Mandatory TrackingAccuracy parameter not present")
+			http.Error(w, "Mandatory TrackingAccuracy parameter not present", http.StatusBadRequest)
+			return
+		}
+	*/
 
-	newSubsId := nextZonalSubscriptionIdAvailable
-	nextZonalSubscriptionIdAvailable++
+	newSubsId := nextAreaCircleSubscriptionIdAvailable
+	nextAreaCircleSubscriptionIdAvailable++
 	subsIdStr := strconv.Itoa(newSubsId)
 	/*
 		if zonalTrafficSub.Duration > 0 {
 			//TODO start a timer mecanism and expire subscription
 		}
 		//else, lasts forever or until subscription is deleted
 	*/
-	if zonalTrafficSub.Duration != 0 { //used to be string -> zonalTrafficSub.Duration != "" && zonalTrafficSub.Duration != "0" {
+	if areaCircleSub.Duration != 0 { //used to be string -> zonalTrafficSub.Duration != "" && zonalTrafficSub.Duration != "0" {
+		//TODO start a timer mecanism and expire subscription
+		log.Info("Non zero duration")
+	}
+	//else, lasts forever or until subscription is deleted
+
+	areaCircleSub.ResourceURL = hostUrl.String() + basePath + "subscriptions/area/circle/" + subsIdStr
+
+	_ = rc.JSONSetEntry(baseKey+typeAreaCircleSubscription+":"+subsIdStr, ".", convertAreaCircleSubscriptionToJson(areaCircleSub))
+
+	registerAreaCircle(areaCircleSub, subsIdStr)
+
+	response.CircleNotificationSubscription = areaCircleSub
+
+	jsonResponse, err := json.Marshal(response)
+	if err != nil {
+		log.Error(err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusCreated)
+	fmt.Fprintf(w, string(jsonResponse))
+}
+
+func areaCircleSubPut(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	vars := mux.Vars(r)
+
+	var response InlineCircleNotificationSubscription
+
+	var body InlineCircleNotificationSubscription
+	decoder := json.NewDecoder(r.Body)
+	err := decoder.Decode(&body)
+	if err != nil {
+		log.Error(err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	areaCircleSub := body.CircleNotificationSubscription
+
+	if areaCircleSub == nil {
+		log.Error("Body not present")
+		http.Error(w, "Body not present", http.StatusBadRequest)
+		return
+	}
+
+	//checking for mandatory properties
+	if areaCircleSub.CallbackReference == nil || areaCircleSub.CallbackReference.NotifyURL == "" {
```

Comment:
This code seems to be duplicated. Can/should we use a function?

Reply:
the code is unique for each method and each endpoint. Some are similar but each have their differences too. I think this is a clear section at the beginning of each function, so i would it as is... The ones that are very close are the POST and PUT for each endpoints, which look the same except to PUT also checking for ResourceUrl being there. I think duplication is then pretty limited. That being said, your comment is welcome if you feel it adds to clarity to create functions for each endpoint method.

Reply:
okay to leave as is, thanks!